### PR TITLE
cl: align blob pruning with configured serve range

### LINKED ISSUE #23752: cl/blob_storage: pruning rejects writes inside the non-archive backfill window
## Problem

Blob cleanup uses a fixed retention distance of 128,600 slots, while the non-archive blob history downloader scans back to `BeaconChainConfig.MinSlotsForBlobsSidecarsRequest()`.

The configured downloader windows are:

- mainnet: `4096 * 32 = 131072` slots
- Gnosis and Chiado: `16384 * 16 = 262144` slots

After cleanup calls `BlobStore.PruneBelow`, `bucketStore.startWrite` rejects every write whose slot is below the pruning floor. `WriteBlobSidecars` returns `nil` for that rejection.

As a result, slots in `[currentSlot - configuredWindow, currentSlot - 128600)` are selected for non-archive backfill but cannot be persisted. The durable post-check finds the sidecars missing and retries them until they age out of the downloader window.

## Impact

- On mainnet, the bottom 2,472 slots of the advertised backfill window cannot be repaired after the pruning floor advances.
- On Gnosis and Chiado, the affected range is 133,544 slots.
- The rejected write returns no error, so logs do not identify the retention mismatch as the cause.
- Immediate blob backfill can remain incomplete while repeatedly fetching data that storage will not accept.

This is present on `main` and is inherited by the release/3.6 backport in #23732.

## Suggested fix

Use one chain-configured retention boundary for both pruning and non-archive backfill. If the protocol window is the intended retention policy, derive the pruning distance from `MinSlotsForBlobsSidecarsRequest()`. If 128,600 slots is intentional, clamp the downloader to that same boundary instead.

Add an integration test that advances the storage pruning floor and then backfills a missing sidecar near the lower edge of the configured downloader window.

Related retention tracking: #23024. Related missing-sidecar consequences: #23224.

### PR DESCRIPTION
**Pull request by domiwei**
Fixes #23752.

## Summary
- derive blob pruning and non-archive backfill floors from the configured wall-clock epoch window
- clamp the mandatory blob range to Deneb and use the PeerDAS window for Fulu-era recovery
- repair count-only holes left by prior pruning without retrying Fulu slots after their data-column window expires
- preserve archive/no-pruning behavior and exact-floor writes

## Operational impact
- mainnet retention increases from 128,600 to 131,072 slots (about 17.9 to 18.2 days at 12-second slots)
- Gnosis and Chiado retention increases from 128,600 to 262,144 slots (about 7.4 to 15.2 days at 5-second slots)

The Gnosis/Chiado change nearly doubles the retained blob horizon; actual disk growth depends on blob usage and storage characteristics.

## Validation
- go test ./cl/clparams ./cl/phase1/network ./cl/phase1/stages -count=1
- focused race tests for pruning, restart recovery, wall-clock boundaries, and Fulu retry aging
- fast and full golangci-lint: 0 issues
- go mod tidy with no diff
- make erigon integration
- adversarial Standards/Spec, Boundary, Lifecycle, and evidence-truth review

### CODE REVIEW

**Comment by AskAlexSharov** [comment]
No blocking correctness bug. The epoch-boundary floor and the Deneb/Fulu clamps are right, and the overflow edges (`SlotsPerEpoch == 0`, unscheduled fork epoch = MaxUint64, `currentEpoch < minEpochs`) are all covered.

What to cut:

1. `cl/clparams/config.go:423`, `cl/phase1/network/blob_downloader.go:429`, `cl/phase1/stages/cleanup_and_pruning.go:54` — three copies of the same formula, `max(forkEpoch, epoch - minEpochs) * slotsPerEpoch`. `specColumnFloor` and `dataColumnServeRangeStartSlot` answer the same question ("first column slot we must serve") and already disagree: the downloader clamps to `FuluForkEpoch`, the pruner does not. Harmless today, but the next edit to one will not reach the other. Replace all three with one helper:

```go
func serveRangeStartSlot(currentSlot, slotsPerEpoch, minEpochs, forkEpoch uint64) uint64 {
	if slotsPerEpoch == 0 {
		return 0
	}
	epoch := currentSlot / slotsPerEpoch
	if epoch < forkEpoch {
		return 0
	}
	return max(forkEpoch, epoch-min(epoch, minEpochs)) * slotsPerEpoch
}
```

`min` gives the saturating subtraction, so the separate `currentEpoch <= minEpochs` branch goes away. Call it with `DenebForkEpoch`, `FuluForkEpoch`, and `0` respectively. ~45 lines -> ~9.

2. `cl/phase1/network/blob_downloader.go:332` — drop `if b.ethClock != nil`. `NewBlobHistoryDownloader` has one call site (`clstages.go:117`) and it always passes a clock, so the branch only exists for tests. The fallback is `retentionSlot = currentSlot` (the stage head) — exactly the bug this PR fixes — and it would fire silently. The new tests already build `eth_clock.NewMockEthereumClock`; give it to the rest and make the field required.

3. `cl/phase1/network/blob_downloader.go:485` and `:895` — `!b.archiveBlobs && block.Version() >= clparams.FuluVersion && slot < fuluRetentionFloor` written twice. One small method on the downloader.

4. `cl/phase1/stages/cleanup_and_pruning_test.go` — `TestCleanupAndPruningKeepsConfiguredBlobRequestWindowWritable` spends ~50 lines on a real `BlobStore` + memfs and two sidecar writes to prove the prune floor. `TestCleanupAndPruningLogsPruneErrors` already pins the exact floor with `PruneBelow(68_928)`, and `bucket_store` already tests that a write below `pruneFloor` is refused. Delete it.

5. `cl/phase1/stages/cleanup_and_pruning_test.go:123` — `beaconCfg.DenebForkEpoch = 0` in `TestCleanupAndPruningKeepsEveryBlobUnderArchiveFlags` has no effect: that path forces `blobFloor = 0` before the config is read.

Worth a line in the PR body: on Gnosis/Chiado `MinEpochsForBlobSidecarsRequests` is 16384 with 16 slots/epoch, so the retained window goes 128600 -> 262144 slots (~7.4 -> ~15.2 days at 5s slots). Spec-correct, but it is a ~2x blob-disk jump for existing non-archive operators on those chains. Mainnet moves 128600 -> 131072, which is noise.

**Comment by domiwei** [comment]
Addressed the simplification points in `4d01c8c901`:

- consolidated the epoch-aligned serve-range arithmetic in `clparams`; both blob and data-column callers now use the same helper, including the Fulu fork clamp
- made `ethClock` a required constructor dependency and removed the stage-head fallback; test fixtures now provide a mock clock
- extracted the shared Fulu retention predicate used by fresh scans and retries
- added the mainnet/Gnosis/Chiado retention-horizon and disk-impact note to the PR body

I kept `TestCleanupAndPruningKeepsConfiguredBlobRequestWindowWritable`: it is the issue's cross-layer regression test proving that cleanup plus the real `BlobStore` accepts the exact configured floor and rejects floor-1. The mock floor assertion and the bucket-store unit tests do not catch those two layers being wired to different boundaries.

I also kept `beaconCfg.DenebForkEpoch = 0` in the archive-flags test. The production code computes the config-derived floor before the archive override. At that fixture's slot, the unmodified mainnet config is pre-Deneb and already returns zero, so removing the assignment would make the test pass even if the archive override stopped working.

Validation: affected package tests, full `make lint`, and `make erigon integration` pass. Independent boundary and lifecycle follow-up reviews converged with no in-scope findings.

File: cl/phase1/network/blob_downloader.go [hunk 1]
```diff
@@ -321,9 +328,12 @@ func (b *BlobHistoryDownloader) downloadOnce(shouldLog bool) error {
...
 	targetSlot := b.nextBackfillTargetSlot
 	retryFloor := uint64(0)
+	fuluRetentionFloor := uint64(0)
 	// in case of non-archive mode, we only backfill the last relevant epochs
 	if !b.archiveBlobs {
-		retentionFloor := currentSlot - min(currentSlot, b.beaconCfg.MinSlotsForBlobsSidecarsRequest())
+		retentionSlot := b.ethClock.GetCurrentSlot()
+		retentionFloor := b.beaconCfg.BlobSidecarServeRangeStartSlot(retentionSlot)
+		fuluRetentionFloor = b.beaconCfg.DataColumnSidecarServeRangeStartSlot(retentionSlot)
 		targetSlot = max(targetSlot, retentionFloor)
```
Comment (awskii) [inline cl/phase1/network/blob_downloader.go]: `currentSlot` is the stage head (`:313`, `b.headSlot.Load()`), but the floor now comes from wall clock. The old floor, `currentSlot - min(currentSlot, MinSlotsForBlobsSidecarsRequest())`, was `<= currentSlot` by construction, so the loop at `:343` always ran at least one iteration. Mixing the two slot sources drops that invariant.

When the head lags wall clock by more than the full serve window, `retentionFloor > currentSlot`, `for currentSlot >= targetSlot` is false on its first evaluation, and control falls through to `:425` — `setBackfillCompleted(true)` whenever `retryRanges` is empty. So the pass reports completion having downloaded nothing. Both `syncedChecker.Synced()` and `retryFailedRecoveries` sit inside that loop, so pending retries neither drain nor get trimmed for as long as the condition holds.

The window is `MinEpochsForBlobSidecarsRequests` — 4096 on mainnet (`config.go:924`), about 18 days. Reachable on a resume rather than only on a stall: `resolveResumeHorizonSlots` permits a local finalized state exactly that stale, and the history download itself adds to the lag.

`targetSlot = max(targetSlot, min(retentionFloor, currentSlot))` restores the old invariant without giving up the wall-clock floor.

File: cl/phase1/network/blob_downloader.go [hunk 2]
```diff
@@ -897,15 +919,13 @@ func (b *BlobHistoryDownloader) collectIncompleteBlocks(currentSlot, targetSlot
 	return batch, visited, nil
 }
 
-// storedSidecarsAvailable reports whether a count-equal slot is actually backed by files. Pruning
-// drops sidecar files but leaves their count rows, and a rewrite can fail partway, so on an archive
-// node a matching count alone is not evidence the store can serve the slot.
-//
-// Non-archive nodes prune on purpose, so a missing file there is expected and must not be re-queued.
+func (b *BlobHistoryDownloader) outsideFuluRetention(block *cltypes.SignedBeaconBlock, slot, retentionFloor uint64) bool {
+	return !b.archiveBlobs && block.Version() >= clparams.FuluVersion && slot < retentionFloor
```
Comment (awskii) [inline cl/phase1/network/blob_downloader.go]: The `!b.archiveBlobs` term is already implied by its own argument. `fuluRetentionFloor` is assigned only inside the `if !b.archiveBlobs` block at `:331-336` and stays `0` otherwise, and `archiveBlobs` is fixed at construction — so on an archive node `retentionFloor == 0` and `slot < 0` is false for every slot regardless.

The cost is not the redundant term but what it invites in the tests: the `archive below floor` subtest at `blob_downloader_boundary_test.go:574` sets `archive: true` with `fuluRetentionFloor: slot + 1` and calls `retryFailedRecoveries(0, slot+1)` directly. No production path can produce that pair, so the subtest pins a state that cannot occur.

Dropping the term leaves `block.Version() >= clparams.FuluVersion && slot < retentionFloor`, and the floor alone carries the archive policy — decided at the one place that already decides it.

File: cl/phase1/network/blob_downloader_boundary_test.go [hunk 3]
```diff
@@ -770,19 +832,34 @@ func TestBlobHistoryDownloaderInteriorResolveKeepsRetryRangeBound(t *testing.T)
...
 
-func TestBlobHistoryDownloaderDropsRetriesBeforeNonArchiveRetentionFloor(t *testing.T) {
-	retention := clparams.MainnetBeaconConfig.MinSlotsForBlobsSidecarsRequest()
-	headSlot := retention + 10
-	expiredSlot := uint64(9)
-	reader := &boundaryBlockReader{}
+func TestBlobHistoryDownloaderUsesWallClockEpochRetentionFloor(t *testing.T) {
+	const (
+		wallClockSlot    = uint64(100)
+		headSlot         = uint64(99)
```
Comment (awskii) [inline cl/phase1/network/blob_downloader_boundary_test.go]: These two slots are in the same epoch: with `SlotsPerEpoch = 16`, `99/16` and `100/16` are both 6, so `serveRangeStartSlot` returns `(6-2)*16 == 64` from either input.

This is the only test covering the change's central behaviour — the floor moving from the stage head to `ethClock.GetCurrentSlot()` — and every assertion in it holds identically if that reverts. The `ErrorIs` on the error seeded at slot 64, the `NotContains(reader.slots, 63)` and the `Contains(reader.slots, 64)` are all satisfied by the head-derived floor too.

The only thing that would catch the regression is gomock's implicit `Times(1)` on `clock.EXPECT().GetCurrentSlot()` — a call-count assertion rather than an observable one. A later refactor that keeps a `GetCurrentSlot()` call for logging or a head check while computing the floor from `headSlot` leaves this green.

Putting the two in different epochs discriminates: `wallClockSlot = 112` (epoch 7, floor 80) against `headSlot = 99` (epoch 6, floor 64).

**Review by awskii — APPROVED** [review]
Checked two of the removals rather than assuming them, and both are sound. The `ArchiveBlobs || BlobPruningDisabled` rewrite is equivalent — the old `floorFor(head, math.MaxUint64)` also returned 0. And dropping `storedSidecarsAvailable`'s `!archiveBlobs` early return cannot re-queue deliberately pruned sidecars, because `firstUnfrozenSlot = max(targetSlot, FrozenBlobs())` with `targetSlot >= retentionFloor` already holds the scan above the pruning floor.

One behaviour change the title does not cover: `serveRangeStartSlot` clamps at `max(forkEpoch, …)` where the old `specColumnFloor` returned 0 for `epoch <= minEpochs`. Early in a fork the floor moves from slot 0 up to the fork boundary. Harmless in practice — no blobs before Deneb, no columns before Fulu — but it is a change in what gets deleted, and the blob window also shifts from a flat 128600 slots to a per-chain `MinEpochsForBlobSidecarsRequests * SlotsPerEpoch`, which is not the same number on every chain. Both are worth a line in the PR body.

Not this PR's work, so no request attached, but worth sequencing: `collectIncompleteBlocks` reads the canonical root and then discards it, returning `[]*cltypes.SignedBeaconBlock`, so the recovery paths re-hash a payload-stripped block. #24053 threads that root through and closes it, and it edits the same function this PR does — the two will conflict.
